### ISSUE TITLE
can't use tokio-tungstenite with tokio-timer

### ISSUE THREAD

**Issue by bbigras** [snapview/tokio-tungstenite#8]
I'm not sure if the problem is with tokio-tungstenite or with tokio-timer.

To reproduce, run `examples/server.rs` with an extra `extern crate tokio_timer;`.

result:
```
> cargo build 
   Compiling test-tung-server v0.1.0 (file:///home/bbigras/dev/rust/tests/test-tung-server)
error[E0283]: type annotations required: cannot resolve `(): std::convert::From<_>`
  --> src/main.rs:77:36
   |
77 |                 let ws_writer = rx.fold(sink, |mut sink, msg| {
   |                                    ^^^^

error: aborting due to previous error

error: Could not compile `test-tung-server`.

To learn more, run the command again with --verbose.
```

rustc 1.17.0-nightly (7846dbe0c 2017-03-26)
futures 0.1.12
tokio-core 0.1.6
tokio-timer 0.1.1
tokio-tungstenite 0.1.2
tungstenite 0.2.1

**Comment by application-developer-DA** [comment]
I guess it's neither `tokio-tungstenite`, nor `tokio-timer`, it's just that sometimes it's really a challenge to figure out where the Rust error comes from. I personally also think that it's quite annoying when such errors occur without a [visible] reason, if you want you can create an issue in `tokio-timer` repository and referenec this issue, maybe we'll get some insights which explain why exactly this code stopped compiling after adding `tokio-timer`. But it seems more like a general problem with the way Rust shows errors, sometimes it's extremely unobvious and the one has to spend a lot of time to figure out what's going on there.

Nevertheless, I found a solution for you which fixes this issue and makes the code compilable again. Here is the diff:

```diff
diff --git a/examples/server.rs b/examples/server.rs
index c757829..998c646 100644
--- a/examples/server.rs
+++ b/examples/server.rs
@@ -21,6 +21,7 @@ extern crate futures;
 extern crate tokio_core;
 extern crate tokio_tungstenite;
 extern crate tungstenite;
+extern crate tokio_timer;
 
 use std::cell::RefCell;
 use std::collections::HashMap;
@@ -28,13 +29,14 @@ use std::env;
 use std::io::{Error, ErrorKind};
 use std::rc::Rc;
 
-use futures::stream::Stream;
-use futures::Future;
-use tokio_core::net::TcpListener;
+use futures::stream::{Stream, SplitSink};
+use futures::{Future, Sink};
+use futures::sync::mpsc::UnboundedSender;
+use tokio_core::net::{TcpListener, TcpStream};
 use tokio_core::reactor::Core;
 use tungstenite::protocol::Message;
 
-use tokio_tungstenite::accept_async;
+use tokio_tungstenite::{accept_async, WebSocketStream};
 
 fn main() {
     let addr = env::args().nth(1).unwrap_or("127.0.0.1:8080".to_string());
@@ -86,15 +88,14 @@ fn main() {
                                 .filter(|&(&k, _)| k != addr)
                                 .map(|(_, v)| v);
                 for tx in iter {
-                    tx.send(message.clone()).unwrap();
+                    (tx as &UnboundedSender<Message>).send(message.clone()).unwrap();
                 }
                 Ok(())
             });
 
             // Whenever we receive a string on the Receiver, we write it to
             // `WriteHalf<WebSocketStream>`.
-            let ws_writer = rx.fold(sink, |mut sink, msg| {
-                use futures::Sink;
+            let ws_writer = rx.fold(sink, |mut sink, msg| -> Result<SplitSink<WebSocketStream<TcpStream>>, ()> {
                 sink.start_send(msg).unwrap();
                 Ok(sink)
             });

```

**Comment by application-developer-DA** [comment]
@bbigras have you tried the workaround? (the server.rs code compiles well with it)

Did it help you to compile the code? (if so, feel free to close the issue)

**Comment by compressed** [comment]
@bbigras this has to do with UnboundedSender and sink both exposing `send` fns. Here's the related issue: https://github.com/alexcrichton/futures-rs/issues/261.

**Comment by application-developer-DA** [comment]
@compressed this particular issue was caused by a `fold()` argument, after adding `tokio-timer` as external crate, Rust cannot make a type inference for the passed closure (don't know why).

I've also noticed the problem with `UnboundedSender` and `Sink` in the past and previously the example workarounded that by importing `futures::Sink` only when the `Sink` trait is actually used. But in my latest diff I decided to rewrite that part with `UnboundedSender::send()` explicitly.
